SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,710 @@
+      // specialization is disabled.
+      //
+      // TODO: A @noescape closure should never be converted to an @owned argument regardless of the function's effect
+      // attribute.
+      if !callee.effectAllowsSpecialization && !onlyHaveThinToThickClosure {
+        continue
+      }
+
+      // Avoid an infinite specialization loop caused by repeated runs of ClosureSpecializer and CapturePropagation.
+      // CapturePropagation propagates constant function-literals. Such function specializations can then be optimized
+      // again by the ClosureSpecializer and so on. This happens if a closure argument is called _and_ referenced in
+      // another closure, which is passed to a recursive call. E.g.
+      //
+      // func foo(_ c: @escaping () -> ()) { 
+      //  c() foo({ c() })
+      // }
+      //
+      // A limit of 2 is good enough and will not be exceed in "regular" optimization scenarios.
+      let closureCallee = rootClosure is PartialApplyInst 
+                          ? (rootClosure as! PartialApplyInst).referencedFunction!
+                          : (rootClosure as! ThinToThickFunctionInst).referencedFunction!
+
+      if closureCallee.specializationLevel > specializationLevelLimit {
+        continue
+      }
+
+      if callSiteMap[pai] == nil {
+        callSiteMap.insert(key: pai, value: CallSite(applySite: pai))
+      }
+
+      intermediateClosureArgDescriptors.append(
+        IntermediateClosureArgDescriptor(callSite: pai, closureArgumentIndex: closureArgumentIndex, 
+                                        parameterInfo: closureParamInfo, 
+                                        reachableExitBBs: Array(callee.blocks.filter { $0.isReachableExitBlock })
+      ))
+    } 
+  }
+  return intermediateClosureArgDescriptors
+}
+
+private func handleConversionsAndReabstractions(for rootClosure: SingleValueInstruction,
+                                                conversionsAndReabstractions: inout ValueWorklist, 
+                                                remainingUses: inout ValueWorklist,
+                                                allDirectAndTransitiveUses: inout [Instruction], 
+                                                _ context: FunctionPassContext) -> Bool 
+{
+  var haveUsedReabstraction = false
+
+  /// The root closure may be a `partial_apply [stack]` and we need to make sure that all `mark_dependence` bases for 
+  /// it will be available in the specialized callee, in case the call site is specialized against this root closure.
+  /// 
+  /// `possibleMarkDependenceBases` keeps track of all potential values that may be used as bases for creating
+  /// `mark_dependence`s for our `onStack` root closure. In the obvious cases these values might be non-trivial closure
+  /// captures (which are always available as function arguments in the specialized callee), but they may also be 
+  /// conversions/reabstractions of the root closure.
```

Comment:
While creating an example I realized that some of the wording in the comment was wrong. I've revised it to -
```
  /// The root closure or an intermediate closure created by reabstracting the root closure may be a `partial_apply
  /// [stack]` and we need to make sure that all `mark_dependence` bases for this `onStack` closure will be available in
  /// the specialized callee, in case the call site is specialized against this root closure.
  ///
  /// `possibleMarkDependenceBases` keeps track of all potential values that may be used as bases for creating
  /// `mark_dependence`s for our `onStack` root/reabstracted closures. For root closures these values are non-trivial
  /// closure captures (which are always available as function arguments in the specialized callee). For reabstracted
  /// closures these values may be the root closure or its conversions (below is a short SIL example representing this
  /// case).
```

And here's a SIL example where a reabstracted closure has a `mark_dependence` on a conversion of the root closure.

```
%someFunction = function_ref @someFunction : $@convention(thin) (Int) -> Int
%rootClosure = partial_apply [callee_guaranteed] %someFunction(%someInt) : $@convention(thin) (Int) -> Int
%noescapeRootClosure = convert_escape_to_noescape %rootClosure : $@callee_guaranteed () -> Int to $@noescape @callee_guaranteed () -> Int
%thunk = function_ref @reabstractionThunk : $@convention(thin) (@noescape @callee_guaranteed () -> Int) -> @out Int
%thunkedRootClosure = partial_apply [callee_guaranteed] [on_stack] %thunk(%noescapeRootClosure) : $@convention(thin) (@noescape @callee_guaranteed () -> Int) -> @out Int
%dependency = mark_dependence %thunkedRootClosure : $@noescape @callee_guaranteed () -> @out Int on %noescapeClosure : $@noescape @callee_guaranteed () -> Int
%takesClosure = function_ref @takesClosure : $@convention(thin) (@owned @noescape @callee_guaranteed () -> @out Int)
%ret = apply %takesClosure(%dependency) : $@convention(thin) (@owned @noescape @callee_guaranteed () -> @out Int)
```